### PR TITLE
python3Packages.tensorflow: Fix numpy==1.20 support

### LINKED ISSUE #128829: Python 37, 38, 39: Tensorflow broken
**Describe the bug**
When trying to add a LSTM-layer to a tensorflow model, I get the following error:
```
NotImplementedError: Cannot convert a symbolic Tensor (lstm/strided_slice:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```
It seems like Tensorflow is not compatible with the version of Numpy included in nixpkgs.

Relevant links:
- https://stackoverflow.com/questions/66207609/notimplementederror-cannot-convert-a-symbolic-tensor-lstm-2-strided-slice0-t/66207610
- https://stackoverflow.com/questions/58479556/notimplementederror-cannot-convert-a-symbolic-tensor-2nd-target0-to-a-numpy
- https://github.com/tensorflow/models/issues/9706
- https://github.com/tensorflow/tensorflow/issues/47691

**To Reproduce**
```py
# example.py
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM
model = Sequential()
model.add(LSTM(256, input_shape=(1, 66), return_sequences=True))
```
```sh
# The same issue can be reproduced with python37, python38, python39. I didn't test older versions.
nix-shell -p "python37.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
nix-shell -p "python38.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
nix-shell -p "python39.withPackages(ps: [ps.tensorflow])" --run "python example.py" --pure
```

**Expected behavior**
Adding the LSTM-layer should not throw an error.

**Traceback**
```
2021-06-30 23:51:17.280747: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
Traceback (most recent call last):
  File "/home/tobias/example.py", line 6, in <module>
    model.add(layer)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/training/tracking/base.py", line 517, in _method_wrapper
    result = method(self, *args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/sequential.py", line 208, in add
    layer(x)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 660, in __call__
    return super(RNN, self).__call__(inputs, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 951, in __call__
    return self._functional_construction_call(inputs, args, kwargs,
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 1090, in _functional_construction_call
    outputs = self._keras_tensor_symbolic_call(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 822, in _keras_tensor_symbolic_call
    return self._infer_output_signature(inputs, args, kwargs, input_masks)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/engine/base_layer.py", line 863, in _infer_output_signature
    outputs = call_fn(inputs, *args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent_v2.py", line 1157, in call
    inputs, initial_state, _ = self._process_inputs(inputs, initial_state, None)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 859, in _process_inputs
    initial_state = self.get_initial_state(inputs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 642, in get_initial_state
    init_state = get_initial_state_fn(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 2506, in get_initial_state
    return list(_generate_zero_filled_state_for_cell(
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 2987, in _generate_zero_filled_state_for_cell
    return _generate_zero_filled_state(batch_size, cell.state_size, dtype)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 3003, in _generate_zero_filled_state
    return nest.map_structure(create_zeros, state_size)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/nest.py", line 659, in map_structure
    structure[0], [func(*x) for x in entries],
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/nest.py", line 659, in <listcomp>
    structure[0], [func(*x) for x in entries],
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/keras/layers/recurrent.py", line 3000, in create_zeros
    return array_ops.zeros(init_state_size, dtype=dtype)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2819, in wrapped
    tensor = fun(*args, **kwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2868, in zeros
    output = _constant_if_small(zero, shape, dtype, name)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/ops/array_ops.py", line 2804, in _constant_if_small
    if np.prod(shape) < 1000:
  File "<__array_function__ internals>", line 5, in prod
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3030, in prod
    return _wrapreduction(a, np.multiply, 'prod', axis, dtype, out,
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 87, in _wrapreduction
    return ufunc.reduce(obj, axis, dtype, out, **passkwargs)
  File "/nix/store/vhcrcs14gd0j618wsas1v8dxnr8z5j6d-python3-3.9.4-env/lib/python3.9/site-packages/tensorflow/python/framework/ops.py", line 852, in __array__
    raise NotImplementedError(
NotImplementedError: Cannot convert a symbolic Tensor (lstm/strided_slice:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```

**Additional context**

**Notify maintainers**

@jyp
@abbradar 

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.41-1-MANJARO, Manjaro Linux, noversion`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.12`
 - channels(tobias): `"nixpkgs-21.11pre294396.3bc8e5cd23b, nixpkgs-afdb5675a18-21.11pre293818.afdb5675a18, pkgs-21.11pre294396.3bc8e5cd23b, pkgs-afdb5675a18-21.11pre293818.afdb5675a18"`
 - nixpkgs: `/home/tobias/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by bergkvist**
###### Motivation for this change
Fixes https://github.com/NixOS/nixpkgs/issues/128829

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -1228,6 +1228,12 @@
     githubId = 251106;
     name = "Daniel Bergey";
   };
+  bergkvist = {
+    email = "tobias@bergkv.ist";
+    github = "bergkvist";
+    githubId = 410028;
+    name = "Tobias Bergkvist";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:1236]: Please make this into a separate commit:

```
maintainers: add bergkvist
```

File: pkgs/development/python-modules/tensorflow/support-numpy-1.20.patch [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ # go/tf-wildcard-import
+ # pylint: disable=wildcard-import
+ from tensorflow.python.ops.gen_array_ops import *
+@@ -2897,7 +2898,7 @@ def matrix_set_diag(
+ 
+ def _constant_if_small(value, shape, dtype, name):
+   try:
+-    if np.prod(shape) < 1000:
++    if math_ops.reduce_prod(shape) < 1000:
+       return constant(value, shape=shape, dtype=dtype, name=name)
```
Comment (jonringer) [inline pkgs/development/python-modules/tensorflow/support-numpy-1.20.patch:19]: I feel like this should probably be upstreamed, there's nothing nixpkgs specific in it.

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 3]
```diff
@@ -335,7 +337,7 @@ let
       description = "Computation using data flow graphs for scalable machine learning";
       homepage = "http://tensorflow.org";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tensorflow/default.nix:338]:
```suggestion
      homepage = "https://tensorflow.org";
```

File: pkgs/development/python-modules/tensorflow/support-numpy-1.20.patch (hunk 2 above)
Comment (bergkvist) [reply pkgs/development/python-modules/tensorflow/support-numpy-1.20.patch:19]: I agree that this fix feels like it belongs in the Tensorflow repository. (or perhaps even in the implementation of np.prod in the Numpy repository)

There have been open PRs to resolve the issue upstream/in the Tensorflow repo for more than 2 months. Supporting Numpy 1.20 doesn't seem to be a high priority for the Tensorflow team - since Numpy 1.19 is working fine with Tensorflow.

That said, Numpy 1.20-support is pretty much essential in order for Tensorflow to work within the nixpkgs ecosystem, since nixpkgs currently ships NumPy 1.20 (pkgs.python3.pkgs.numpy).

Shipping multiple versions of the same Python library within nixpkgs is a risky affair - since any environment that depends on multiple versions of the same library will break.

**Comment by risicle** [comment]
Upstream seem to have fixed this via https://github.com/tensorflow/tensorflow/pull/47957

**Comment by bergkvist** [comment]
@risicle Okay - nice! Does that means we should close this pull request? (Once we verify that Tensorflow in nixpkgs is working)

I guess an alternative would be to modify this PR to bump the Tensorflow version instead of patching it with the fix.

**Comment by risicle** [comment]
Either works for me.

**Comment by risicle** [comment]
Fixed in #132099
